pkg/arvo/app/http-server-store.hoon

```diff
@@ -0,0 +1,220 @@
+=*  state  -
+^-  agent:gall
+|_  =bowl:gall
++*  this  .
+    def   ~(. (default-agent this %|) bowl)
+::
+++  on-init
+  ^-  (quip card _this)
+  |^
+  :_  %_  this
+          serving
+        %-  ~(gas by *^serving)
+        %+  turn
+          ^-  (list path)
+          [/ /'~landscape' /'~chat' /'~link' /'~dojo' /'~groups' /'~publish' ~]
+        |=(pax=path [pax [/app/landscape %.n]])
+      ==
+  :~  (connect /)
+      (connect /'~landscape')
+      (connect /'~chat')
+      (connect /'~link')
+      (connect /'~dojo')
+      (connect /'~groups')
+      (connect /'~publish')
+  ==
+  ::
+  ++  connect
+    |=  =path
+    ^-  card
+    [%pass path %arvo %e %connect [~ path] %http-server-store]
+  --
+::
+++  on-save  !>(state)
+++  on-load
+  |=  old-vase=vase
+  ^-  (quip card _this)
+  [~ this(state !<(state-zero old-vase))]
+::
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  |^
+  ?>  (team:title our.bowl src.bowl)
+  ?+  mark  (on-poke:def mark vase)
+      %http-server-action  (landscape-action !<(action vase))
+      %handle-http-request
+    =+  !<([id=@ta req=inbound-request:eyre] vase)
+    :_  this
+    %+  give-simple-payload:app  id
+    (handle-http-request req)
+  ==
+  ::
+  ++  landscape-action
+    |=  act=action
+    ^-  (quip card _this)
+    |^
+    ?-  -.act
+        %serve-dir
+      =*  url-base  url-base.act
+      ?:  (~(has by serving) url-base)
+        ~|("url already bound to {<(~(got by serving) url-base.act)>}" !!)
+      :-  [%pass url-base %arvo %e %connect [~ url-base] %http-server-store]~
+      this(serving (~(put by serving) url-base [clay-base.act public.act]))
+    ::
+        %unserve-dir
+      :-  [%pass url-base.act %arvo %e %disconnect [~ url-base.act]]~
+      this(serving (~(del by serving) url-base.act))
+    ::
+        %toggle-permission
+      ?.  (~(has by serving) url-base.act)
+        ~|("url is not bound" !!)
+      =/  [clay-base=path public=?]  (~(got by serving) url-base.act)
+      :-  ~
+      this(serving (~(put by serving) url-base.act [clay-base !public]))
+    ::
+        %set-landscape-homepage-prefix
+      =.  landscape-homepage-prefix.configuration  prefix.act
+      :_  this
+      (give [%configuration configuration])
+    ==
+    ::
+    ++  give
+      |=  =update
+      ^-  (list card)
+      [%give %fact [/all]~ [%http-server-update !>(update)]]~
+    --
+  ::
+  ++  handle-http-request
+    |=  =inbound-request:eyre
+    ^-  simple-payload:http
+    |^
+    =*  req       request.inbound-request
+    =*  headers   header-list.req
+    =/  req-line  (parse-request-line url.req)
+    ?.  =(method.req %'GET')  not-found:gen
+    =?  req-line  ?=(~ ext.req-line)
+      [[[~ %html] ~['index']] args.req-line]
+    ?>  ?=(^ ext.req-line)
+    ?~  site.req-line
+      not-found:gen
+    =*  url-prefix  landscape-homepage-prefix.configuration
+    ?.  ?|  ?&(?=(^ url-prefix) =((need url-prefix) i.site.req-line))
+            =(url-prefix ~)
+        ==
```

Comment:
The specificity is good imo. Normally I wouldn't like it, but I think it's positive here.

Reply:
I'd break it up as
```
?~  landscape-homepage-prefix.configuration
  not-found:gen
?.  =(u.landscape-homepage-prefix.configuration i.site.req-line)
  not-found:gen
```
since the actual content of the if is pretty trivial. (And ban six-syllable 25-character field names so things don't get this awkward.)

Reply:
…hm okay that doesn't work because `~` is the found case, `?. ?~(thing & =(u.thing req-line))`? In which case the fall makes sense ignore me ^-^

Reply:
The middle ground I originally crossed over, and that @ohAitch almost arrived at, is just to do the `~` case check first, so you can do the value comparison more easily:

```hoon
?|  ?=(~ url-prefix)
    =(u.url-prefix i.site.req-line)
==
```

Reply:
Thanks, good suggestion.